File: morning_exercises/day11.ts
class BSTNode {
    constructor (
        public val: number = 0,
        public left: BSTNode | null = null,
        public right: BSTNode | null = null
    ) {}

    static lca(root: BSTNode | null, val1: number, val2: number): number {
        while(root) {
            if (val1<root.val && val2<root.val) root = root.left;
            else if (val1>root.val && val2>root.val) root = root.right;
            else return root.val;
        }
        return 0;
    }
}

{
    const root = new BSTNode(5);
    root.left = new BSTNode(4);
    root.right = new BSTNode(7);
    root.right.left = new BSTNode(6);
    root.right.right = new BSTNode(8);
    root.right.right.right = new BSTNode(10);

    console.log(BSTNode.lca(root, 6,8));
}

File: morning_exercises/day11.cs
public class BSTNode(int val = 0, BSTNode? left = null, BSTNode? right = null)
{
    public int val = val;
    public BSTNode? left = left;
    public BSTNode? right = right;

    public static int Lca(BSTNode? root, int val1, int val2)
    {
        while (root!=null)
        {
            if ((root.val > val1) && (root.val > val2))
            {
                root = root.left;
            }
            else if ((root.val < val1) && (root.val < val2))
            {
                root = root.right;
            }
            else return root.val;
        }
        return 0; // unreachable — p and q are guaranteed to exist. just keping it here cuz for c# all code paths must return val
    }
}

